Add idempotency and import/export support to zabbix_template

logan2211 opened the pull request:
Adds idempotency to the template update functions and check mode,
also adds the ability to dump and import json template
configurations.

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Rewrote most of the module adding the ability to import full template configurations, including items, graphs, triggers, etc. Additionally the module functions are now idempotent and have better support for check mode.

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
zabbix_template
##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.4.1.0
  config file = None
  configured module search path = [u'/home/ubuntu/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

logan2211 committed:
Add idempotency and import/export support to zabbix_template

Adds idempotency to the template update functions and check mode,
also adds the ability to dump and import json template
configurations.

eikef commented:
Generally, this looks good -- however I'm uncertain that state: dump belongs in a module instead of a lookup plugin. Maybe more seasoned ansible developers can chime in on this?

Logan2211 commented:
I think it makes more sense to keep it here. The use case is the same as `slurp` or any of the `_facts` modules which pull data from an API or system running state, and then register it to a variable for further manipulation. You would be duplicating a lot of effort by having to use and maintain `zabbix_template` functionality (a very specific thing) in 2 different places needlessly.

Logan2211 commented:
Also a lookup would remove the ability to execute the dump API call from a remote host which would break my usage of the module and would not align with any of the other zabbix API modules implemented.

eikef commented on lib/ansible/modules/monitoring/zabbix/zabbix_template.py:
```diff
@@ -244,32 +447,49 @@ def main():
+    elif state == "dump":
+        if not template_ids:
+            module.fail_json(msg='Template not found: %s' % template_name)
+        module.exit_json(changed=True, msg=existing_template_json)
```
Why is changed=True here? Nothing should have changed on Zabbix's side...

eikef replied:
It may make sense to put the dump into a variable of its own instead of msg (which ordinarily contains user-readable messages). Maybe template_json?

logan2211 replied:
Done. Also I changed the return variable to `template_json` instead of `msg` so the dump is named more sensibly in the return.

_edit_: also open to other suggestions than `template_json` if there is something more appropriate.. `msg` just didn't make much sense.

eikef replied:
data is too generic, and template_json was the first thing that came to mind. I'm not married to it though.

eikef commented:
Yeah, I see the points about code-duplication and non-remotability.

If in the future it is decided that is the way to go the functionality can still be transplanted. So keep it as is for now :)

eikef commented:
shipit

eikef commented:
You may want to add yourself to the maintainers of this module as well (in .github/BOTMETA.yml), though if you do that in this commit it will impede automerge if another namespace maintainer puts their shipit in.

eikef commented:
@D3DeFi would you mind taking a look as well? :-) Thanks

Logan2211 commented:
Sorry -- pushed one more thing. A documentation for the `template_json` return value when `state: dump` is used.

eikef commented:
shipit

eikef commented:
shipit

Logan2211 commented:
Thanks for the quick reviews and sorry for requiring another one :(

Small optimization.. comparing using sets instead of sorted list in the idempotency check for macros/groups parameters

logan2211 committed:
Fix issue clearing groups from template

When an empty list is provided for group names, all groups associations
should be cleared from the template. Previous behavior caused the
template to be associated to all existing groups if an empty list
was provided.

eikef commented:
shipit

D3DeFi commented:
Hi, haven't had much time to test this extensively, but basic operations seem to work good :)

I've, by mistake, stumbled upon situation where I wanted to add macro to non existing template and got this error:
```bash
Traceback (most recent call last):                                                                                                                                                                                                                                               File "/tmp/user/0/ansible_z2zydb/ansible_module_zabbix_template.py", line 523, in <module>
    main()
  File "/tmp/user/0/ansible_z2zydb/ansible_module_zabbix_template.py", line 501, in main
    if existing_template_json:
UnboundLocalError: local variable 'existing_template_json' referenced before assignment
```

Maybe some kind of check would be appropriate and then `module.fail_json(msg='Modifying non-existing template')` instead of raw traceback. 


This may be a bit out of scope for module's EXAMPLES section, but I would like to see very minimal template_json example as well. Some kind of JSON struct introducing maybe one item and one trigger? :) 

I would suggest to at least add link to [Zabbix documentation](https://www.zabbix.com/documentation/3.0/manual/api/reference/template/create), but in this case it is not very helpful for building up an appropriate JSON struct.

logan2211 committed:
Fix undefined variable references

Logan2211 commented:
Great catch on the `existing_template_json` issue. It should not fail, it should just create a blank template with any group mappings, macros, and links you defined. Should be fixed with 88999d5d43dc656afa85a3348d136c8db522a94f.

I will work on adding an example. I suppose the example I used for the `dump` return value will work fine?

As for the docs link.. that's a good idea. Do you know if there is a way to get an unpinned docs URL like `latest` for Zabbix? I see the one you pasted is pinned to 3.0 docs.

D3DeFi commented:
Return value from `dump` operation should be good enough as an example. 

Regarding the docs link. I unfortunately don't know about any unpinned version. Maybe something uglier like `https://www.zabbix.com/documentation/<your-zbx-version>/manual/api/reference/template/create` will do the trick.

logan2211 committed:
Add example importing template from ansible variable

Document a sample template import with bare minimum structure.
No items or graphs are added, only 1 application is added to the
template.

Logan2211 commented:
Added a sample template import. I did not add the documentation because I can't find anything relevant in the docs. Even https://www.zabbix.com/documentation/3.0/manual/api/reference/configuration/import is virtually useless for someone trying to determine the json/xml schema.

Personally what I do is create my template in a test environment and then use dump to get a json output for source control.

D3DeFi commented:
Looks good. Ansibot will probably not count my shipit as I am not maintainer of the whole zabbix folder, but here it is anyways :)

D3DeFi commented:
shipit

eikef commented:
shipit

eikef commented:
bot_status

eikef commented:
@d3defi maintainer in same namespace works as well :)

Thanks, @logan2211
